### PR TITLE
Sink 2d

### PR DESCRIPTION
**Pull request by wangguan1995**
### PR types

New features, Bug fixes

### PR changes

- [ppsci/constraint/integral_constraint.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-53089bb461bec591a80c2e193b3cfcac3bc708666536b92de1f43ac876e6e21f) : add the function of integral plane translation
- [ppsci/constraint/interior_constraint.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-5d2475659830dcf186b6bad45a7d2f1bb424cf3a93c26002a32caeab6ff44fb3) : remove sdf pop.
- [ppsci/equation/pde/advection_diffusion.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-f6d41034ac9b61914fad4382f039ca8b8351e3c0fe818dae5dec2cf4db6618ef) : add advection-diffusion equation
- [ppsci/equation/pde/grad_normal.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-6c3f6e3cbcfeaff8198eb48a575300cc5f6ae122178357327c3116f53e99fcb8) : add normal gradients equation for adiabatic boundary condition.
- [ppsci/equation/pde/navier_stokes.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-74eae158d2bd01898b71b21cecc25d6205b85fe0c5c7c1f735b5cecea892fe93): Bug fix for dynamic viscosity, add sympy expression for momentum_x and _y.
- [ppsci/equation/pde/turbulence_zero_eq.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-2453253507ecec9c71803e1182761d15bed466e377b748399fa5550d3003733a) : add zero equation model for turbulence flow.
- [ppsci/geometry/csg.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-37e38a6b4b3646ca8eb8a5fff0e1ee0343c6ea83d6ac7c0f910c171bb2dc0f54) : Bug fix for sampling points, add sdf function.
- [ppsci/geometry/geometry.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-992c327ec0d23c31a256afda441d5990980eb847478c2f1719a9a220ca0d98c9): Add sdf gradient function, add geometry union, add boundary area(weight) calculation
- [ppsci/geometry/geometry_2d.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-26706c4702a6514053d9fdcd659407fe0e3be93c683efae85749ffb070d78d5d): Bug fix for rectangle, add approximate area calculation, add 2d channel geometry for channel flow(sdf sample area normal, etc), add line geometry(sdf sample area normal, etc).
- [ppsci/metric/l2_rel.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-4fbe507ecdab517e4a35aaa0c92226c1049181e8a7dea937ab9ffa3448bfa046) : Bug fix for scalar weight.
- [ppsci/solver/eval.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-f707ffe5b13ce9f51cb34df6bd92ca2d28ebac7f4ef85bfd3e2a7635291a9bc9) : add eval summation for multi validators and write to *.csv file for better records.
- [ppsci/solver/solver.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-819ca840c5fadc12bf4443bdb48adf0be30e146c21ee12ddf8cef6944f0501ff) : Bug fix for no "keep_batch" attributes.
- [ppsci/utils/expression.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-f37f6ff7296b9bd7c0f54f68599c5c5ea2b92ddc2b24319833dc839aed3a2f9a): Perfection of forward and cover the case without validator loss.
- [ppsci/utils/paddle_printer.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-f802fe2fceac1b8b03f69d0e30c3cc8a7b28fbaae8c86480588937c84927c8ab): Add paddle printer for recursive derivation of equations and substitute with sdf_derivatives.
- [ppsci/utils/reader.py](https://github.com/PaddlePaddle/PaddleScience/compare/develop...wangguan1995:sink_2d?expand=1#diff-6a973bd7fedd9b67d9ae3405cc12edae5f5b55f48b61101846239c93f77ea417): Perfection for covering steady-state cases, add input patch for sdf, sdf derivatives, area and normal.

### Describe

- Add first heat transfer case of 3-fins 2D sink (Advection-Diffusion).
- Complement and add traits like sdf, geometry, sampling, geometry bool, sdf derivative, symbol function, etc.
- Important Bug fix for N-S equation.
- New equations: GradNormal, AdvectionDiffusion, ZeroEquation.
- Doc will be submit in next PR.

### CODE REVIEW

File: ppsci/validate/sup_validator.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ class SupervisedValidator(base.Validator):
     def __init__(
         self,
         dataloader_cfg: Dict[str, Any],
-        loss: loss.Loss,
+        loss: Optional[loss.Loss] = None,
```
Comment (HydrogenSulfate) [inline ppsci/validate/sup_validator.py:64]: 改回去

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: ppsci/constraint/interior_constraint.py [hunk 2]
```diff
@@ -163,8 +163,6 @@ def __init__(
                 else:
                     raise NotImplementedError(f"type of {type(value)} is invalid yet.")
 
-        if "sdf" in input:
-            input.pop("sdf")
```
[the remark this answers is not in the document]
Comment (wangguan1995) [reply ppsci/constraint/interior_constraint.py:167]: > 不需要删除这行代码

删除无法获取sdf进行计算

File: ppsci/geometry/geometry.py [hunk 3]
```diff
@@ -95,14 +109,33 @@ def sample_interior(self, n, random="pseudo", criteria=None, evenly=False):
                     "please check correctness of geometry and given creteria."
                 )
 
-        # if sdf_func added, return x_dict and sdf_dict, else, only return the x_dict
+        # if sdf_fun/sdf_grad/area added,
+        # return x_dict and sdf_dict/sdf_grad_dict/area_dict, else, only return the x_dict
         if hasattr(self, "sdf_func"):
```
[the remark this answers is not in the document]
Comment (wangguan1995) [reply ppsci/geometry/geometry.py]: 不是所有的几何都有sdf_func, 如果不判断，会有部分CI挂掉

File: ppsci/geometry/geometry.py [hunk 4]
```diff
@@ -63,6 +63,20 @@ def uniform_points(self, n: int, boundary=True):
         )
         return self.random_points(n)
 
+    def sdf_func(self, points: np.ndarray) -> np.ndarray:
+        pass
```
[the remark this answers is not in the document]
Comment (wangguan1995) [reply ppsci/geometry/geometry.py]: 需要去掉sdf_func基类，通过hasattr判断sdf是否计算

**Pull request closed without merging**